### PR TITLE
Filename: Convert extension-related code to use D slices

### PR DESCRIPTION
**Pull request by Geod24**
Better reviewed commit-by commit.
Spin off #8575

### CODE REVIEW

File: src/dmd/root/filename.d [hunk 1]
```diff
@@ -640,20 +640,34 @@ nothrow:
...
+    }
+
+    /// Ditto
+    extern (D) static int exists(const(char)[] name)
     {
         version (Posix)
         {
             stat_t st;
-            if (stat(name, &st) < 0)
+            if (name.toCStringThen!((v) => stat(v.ptr, &st)) < 0)
```
Comment (jacob-carlborg) [inline src/dmd/root/filename.d]: The parentheses around `v` in the lambda parameter list is not necessary.

File: src/dmd/root/filename.d [hunk 2]
```diff
@@ -511,16 +511,32 @@ nothrow:
...
     {
-        const(char)* e = FileName.ext(name);
-        if (!e && !ext)
+        return equalsExt(name.toDString, ext.toDString);
+    }
+
+    /// Ditto
+    extern (D) static bool equalsExt(const(char)[] name, const(char)[] ext) pure
+    {
+        auto e = FileName.ext(name);
```
Comment (jacob-carlborg) [inline src/dmd/root/filename.d]: `const`?

**Comment by jacob-carlborg** [comment]
LGTM.

**Comment by Geod24** [comment]
Addressed @jacob-carlborg 's comments

**Review by jacob-carlborg — APPROVED** [review]

**Comment by Geod24** [comment]
This isn't needed anymore, the main PR was merged.

**Pull request closed without merging**